pandas/io/formats/excel.py

```diff
@@ -588,17 +592,11 @@ def _format_header_regular(self):
                     coloffset = len(self.df.index[0])
 
             colnames = self.columns
-            if has_aliases:
-                # pandas\io\formats\excel.py:593: error: Argument 1 to "len"
-                # has incompatible type "Union[Sequence[Optional[Hashable]],
-                # bool]"; expected "Sized"  [arg-type]
-                if len(self.header) != len(self.columns):  # type: ignore[arg-type]
-                    # pandas\io\formats\excel.py:602: error: Argument 1 to
-                    # "len" has incompatible type
-                    # "Union[Sequence[Optional[Hashable]], bool]"; expected
-                    # "Sized"  [arg-type]
+            if self._has_aliases:
+                assert isinstance(self.header, Sized)
```

Comment:
glad to see this.  for some reason my local mypy diagrees with the CI about whether its L601 or L602 that has the problem

Reply:
It seems that the problem was with len(self.header), so it should be line 602, but presumably one line was added on top since then.

Reply:
once we add a runtime isinstance, this sort of duplicates the _has_aliases property.

since self.header is a sequence or bool, maybe _has_aliases could reflect that instead of the explicit checks for the 4 classes.

if we have a guarantee from the is_* call, a cast would be more appropriate than the assert.

Reply:
Updated.
Used casting instead of asserting.
In ``_has_aliases`` check if not boolean rather than 4 sequence-like classes.
Is that what you meant?

Reply:
If the types are correct for the header parameter of ExcelFormatter then this should be fine.  

also cast is a no-op, so no harm using `self.header = cast(Sequence, self.header)` and the other self.header -> header changes wouldn't be needed.

Reply:
OK, changed the casting as you suggested.